File: 백준/Silver/10844. 쉬운 계단 수/쉬운 계단 수.java
import java.util.Scanner;

public class Main {

    static final int maxN = 100;
    static final int maxIdx = 9;
    static final int mod = (int) 1e9;
    static int N;

    // first: stage 단계, second: 위치
    static int[][] dp = new int[maxN][maxIdx + 1];
    static int[] dI = { -1, 1 };

    public static int solution(int stage, int idx) {
        if (stage < 0) return 0;
        if (dp[stage][idx] != 0) return dp[stage][idx];

        for (int k = 0; k < 2; k++) {
            int nextIdx = idx + dI[k];
            if (nextIdx < 0 || nextIdx > maxIdx) continue;
            dp[stage][idx] = (dp[stage][idx] + solution(stage - 1, nextIdx)) % mod;
        }
        return dp[stage][idx];
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        for (int i = 1; i <= maxIdx; i++) {
            dp[0][i] = 1;
        }

        N = sc.nextInt();
        N--;

        int result = 0;
        for (int i = 0; i <= maxIdx; i++) {
            result = (result + solution(N, i)) % mod; 
        }

        System.out.println(result);
        sc.close();
    }
}

File: 백준/Silver/10844. 쉬운 계단 수/쉬운 계단 수.cc
#include <bits/stdc++.h>
using namespace std;

const int maxN = 100;
const int maxIdx = 9;
const int mod = 1e9;
int N;

// first: stage 단계, second: 위치
int dp[maxN][maxIdx + 1];
int dI[] = { -1, 1 };

int solution(int stage, int idx) {
    if (stage < 0) return 0;
    if (dp[stage][idx]) return dp[stage][idx];

    for (int k = 0; k < 2; k++) {
        int nextIdx = idx + dI[k];
        if (nextIdx < 0 || nextIdx > maxIdx) continue;
        dp[stage][idx] = (dp[stage][idx] + solution(stage - 1, nextIdx)) % mod; // 모듈러 연산 추가
    }
    return dp[stage][idx];
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    for (int i = 1; i <= maxIdx; i++) { 
        dp[0][i] = 1;
    }

    cin >> N;
    N--;

    int result = 0;
    for (int i = 0; i <= maxIdx; i++) {
        result = (result + solution(N, i)) % mod; 
    }
    cout << result;

    return 0;
}